File: src/com/anathema_roguelike/entities/characters/foes/traits/StatTrait.java
package com.anathema_roguelike.entities.characters.foes.traits;

import java.util.Optional;

import com.anathema_roguelike.entities.characters.Character;
import com.anathema_roguelike.entities.characters.perks.Buff;
import com.anathema_roguelike.entities.characters.perks.PassivePerk;
import com.anathema_roguelike.stats.characterstats.CharacterStat;
import com.anathema_roguelike.stats.effects.AdditiveCalculation;
import com.anathema_roguelike.stats.effects.Modifier;
import com.anathema_roguelike.stats.effects.MultiplicativeCalculation;

public class StatTrait<T extends CharacterStat> extends Trait<PassivePerk> {
	
	private Class<T> stat;
	private String name;
	private AdditiveCalculation additiveCalculation;
	private MultiplicativeCalculation multiplicativeCalculation;
	
	public StatTrait(String name, Class<T> stat, AdditiveCalculation additiveCalculation) {
		super();
		
		this.stat = stat;
		this.name = name;
		this.additiveCalculation = additiveCalculation;
	}
	
	public StatTrait(String name, Class<T> stat, MultiplicativeCalculation multiplicativeCalculation) {
		super();
		
		this.stat = stat;
		this.name = name;
		this.multiplicativeCalculation = multiplicativeCalculation;
	}
	
	public Class<T> getStat() {
		return stat;
	}

	@Override
	protected PassivePerk createPerk() {
		return new PassivePerk(name) {

			@Override
			public Optional<Buff> getEffect() {
				if(additiveCalculation != null) {
					return Optional.<Buff>of(new Buff(this, new Modifier<>(stat, additiveCalculation)));
				} else {
					return Optional.<Buff>of(new Buff(this, new Modifier<>(stat, multiplicativeCalculation)));
				}
			}
		};
	}
}

File: src/com/anathema_roguelike/entities/characters/foes/traits/StatTrait.scala
/*******************************************************************************
 * Copyright (c) 2019. Carl Minden
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/

package com.anathema_roguelike
package entities.characters.foes.traits

import com.anathema_roguelike.entities.characters.perks.{Buff, PassivePerk, PassthroughPerk}
import com.anathema_roguelike.stats.Stat.CharacterStat
import com.anathema_roguelike.stats.effects.{Calculation, Modifier}

import scala.reflect.runtime.universe._

class StatTrait[T <: CharacterStat : TypeTag](name: String, calculation: Calculation) extends PassthroughPerk[PassivePerk] {

  def getStat: Class[T] = typeTagToClass

  override protected def createPerk: PassivePerk = new PassivePerk(name) {
    override def getEffect: Option[Buff] = {
      new Buff(this, List(new Modifier[T](calculation)))
    }
  }
}